### DISCUSSION TITLE
Enhancement for KPOINTSline() Function

### DISCUSSION THREAD

**Discussion by bmondal94** [irreducible-representations/irrep#72] [General]
Dear Admin,

We have a BandUPpy code that uses irrep for band unfolding from supercell calculations. I want to bring to your attention a small point that, while not necessarily significant, could be worth consideration. 

Within irrep, there is a `KPOINTSline()` function in the `Bandstructure` class. When converting k-points to Cartesian coordinates while dealing with a supercell, this function currently uses the **supercell reciprocal lattice**:
```
KPcart = np.array(kpred).dot(self.RecLattice)
```
However, I believe the correct approach would be to use the **reference primitive cell** rather than the supercell reciprocal lattice for this conversion. **This is because our unfolded bands are on the reference primitive cell.**

One suggestion for modifying the function is as follows:
```
def KPOINTSline(self, kpred=None, breakTHRESH=0.1, supercell=None):
      '''
     supercell : array, shape=(3,3)
            Describes how the lattice vectors of the (super)cell used in the 
            calculation are expressed in the basis vectors of the primitive 
            cell.
      '''
       if supercell is None:
          reciprocal_lattice = self.RecLattice
      else:
          # Similar to 'unfold()' function within 'Kpoints' class.
          reciprocal_lattice = np.dot(self.RecLattice, np.linalg.inv(supercell.T)) <== Needs to be reviewed properly
      if kpred is None:
            kpred = [k.K for k in self.kpoints]
      KPcart = np.array(kpred).dot(reciprocal_lattice)
```
This adjustment ensures the use of the reference primitive cell for conversion, aligning with the context of the unfolded bands as well.

Best Regards,
Badal

**Comment by stepan-tsirkin** [comment]
Hi @bmondal94,

Yes, such change makes sense, you are welcome to make a small PR for that. Sure, it will not break any of the previous functionalities inside irrep

Some suggestions:
1) As I understand, this makes sense only when kpred is provided (in the PC reciprocal vectors), because otherwise bothe kpred and RecLattice are consistently in the SCBZ . So add

```python
assert kpred is not None
```

2) I think the recalculation shpould be different:

RealLatticeSC = supercel@RealLatticePC

RecLatticePC = 2pi * inv(RealLatticePC).T

RecLatticeSC = 2pi * inv(RealLatticeSC).T = 2pi*inv(supercel@RealLatticePC).T = 2*pi*(inv(RealLatticePC) @ inv(supercell) ).T = 2*pi*inv(supercell.T)  @ inv(RealLatticePC).T =  inv(supercell.T) @ RecLatticePC

Therefore 
RecLatticePC = supercell.T @ RecLatticeSC

so the line should read 

``` python
 if supercell is None:
          reciprocal_lattice = self.RecLattice
      else:
          # Similar to 'unfold()' function within 'Kpoints' class.
          reciprocal_lattice =supercell.T @ self.RecLattice <== Needs to be reviewed properly
```

Do you agree?  

Anyway, you may check it for some system and see if the results make sense

**Reply by bmondal94** [reply]
Ha Ha...
Yes, you are right. I miss printed the reciprocal_lattice conversion line in my previous msg completely. 

I implemented it for banduppy. Although..., I now noticed that I did a mistake in my math any way. I did the matrix multiplication in the opposite order. I did
`reciprocal_lattice_pc = reciprocal_lattice_sc @ supercell.T`
when it should be 
`reciprocal_lattice_pc = supercell.T @ reciprocal_lattice_sc`

Thanks a lot.

And I will create a PR with the suggested improved function here.
